extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -16,12 +17,127 @@ import {
+ * Validates url of shopping config attributes
+ * @param {?string} field
+ * @param {?string} url
+ * @private
+ * @return {!Array<string>}
+ */
+function validateURL(field, url) {
+  if (url == null) {
+    /* Not a Required Attribute, return.*/
+    return;
+  }
+
+  const urls = Array.isArray(url) ? url : [url];
+
+  return urls.map((url) => {
+    try {
+      assertHttpsUrl(url, 'amp-story-shopping-config ' + field);
+    } catch (e) {
+      return e.message;
+    }
+  });
+}
 export class AmpStoryShoppingConfig extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
     /** @private @const {?../../amp-story/1.0/amp-story-store-service.AmpStoryStoreService} */
     this.storeService_ = null;
+
+    /** @private @const {!Object<Array<Function>>} */
+    this.productValidationConfig_ = {
+      /* Required Attrs */
+      'productId': [validateRequired, validateStringLength], // Required. String. Keys to amp-story-shopping-tag nodes.
+      'brandLabel': [validateRequired, validateStringLength], // Required. String.
+      'productTitle': [validateRequired, validateStringLength], // Required. String.
+      'productPrice': [validateRequired, validateNumber], // Required. Number.
+      'productImages': [validateRequired, validateURL], // Required. String or array of strings.
+      'productDetails': [validateRequired, validateStringLength], // Required. String.
+      'reviewsPage': [validateRequired, validateURL], // Required if reviews-data. Links to a page where reviews can be read.
+      'productPriceCurrency': [validateRequired, validateStringLength], // Required. String. ISO 4217 currency code used to display the correct currency symbol.
+      /* Optional Attrs */
+      'productColor': [validateStringLength], // Optional. String.
+      'productSize': [validateStringLength], // Optional. String.
+      'productIcon': [validateURL], // Optional. Links to an image. Defaults to a shopping bag icon.
+      'productTagText': [validateStringLength], // Optional. String.
+      'reviewsData': [validateURL], // Optional. Links to review data.
+      'ctaText': [validateNumber], // Optional. Number. Defaults to “Buy now”. Keys to a CTA text option for i18n.
+      'shippingText': [validateNumber], // Optional. Number. Keys to a shipping text option for i18n. Shipping text will not be rendered if the string is omitted.
+    };
+  }
+  /**
+   * Validates shopping config.
+   * @param {!ShoppingConfigDataDef} shoppingConfig
+   * @private
+   */
+  validateConfig_(shoppingConfig) {
+    const errors = Object.keys(this.productValidationConfig_)
+      .reduce((errors, k) => {
+        const value = shoppingConfig[k];
+        const fns = this.productValidationConfig_[k];
+        errors.push(
+          ...fns.reduce((errors, fn) => {
+            const localErrors = fn(k, value) || [];
+            errors.push(...localErrors);
+            return errors;
+          }, [])
+        );
```

Comment:
You only need the previous `.reduce` accumulator, not this one:

```suggestion
        this.productValidationConfig_[k].forEach(fn => {
          errors.push(...(fn(k, value) || []));
        });
```